SHARD-1860: block eip2930 txs with > a set number of addresses in accessList or storage keys per address

ahmxdiqbal opened the pull request.

ahmxdiqbal committed:
block eip2930 txs with > a set number of addresses in accessList or storage keys per address

BelfordZ commented:
After rebasing you should see no issue with CI

mhanson-github commented on src/index.ts:
```diff
@@ -5210,8 +5210,16 @@ const shardusSetup = (): void => {
 
-        if (isEIP2930 && isStakingEVMTx(transaction) === false) {
-          return { status: false, reason: `EVM txs have been disabled` }
+        const tooManyAddresses = (transaction as AccessListEIP2930Transaction).AccessListJSON?.length > ShardeumFlags.accessListSizeLimit;
```
Is there any situation where (transaction as AccessListEIP2930Transaction) will return null?  causing the .AccessListJson to throw?

Also minor thing but doing the cast 2x in a row for the same thing.

ahmxdiqbal replied:
(transaction as AccessListEIP2930Transaction) doesnt return a value, it just tells the compiler what type to treat transaction as. The only situation that i can see where it would be null is if transaction itself was null. looking over the code above it, we will throw an error before we even get to this if that is the case

ahmxdiqbal replied:
avoided doing the cast twice in latest commit

mhanson-github commented on src/index.ts:
```diff
@@ -5210,8 +5210,16 @@ const shardusSetup = (): void => {
+          (accessListItem) => accessListItem.storageKeys?.length > ShardeumFlags.accessListSizeLimit
+        );
+        
+        if (isEIP2930 && (tooManyAddresses || tooManyStorageKeys)) {
```
Should all of the changes be inside an if (isEIP2930) check?  (IE we don't even need to do the const tooManyStorageKeys/etc. parts if the if check is false?

ahmxdiqbal replied:
yeah we could do that as well

ahmxdiqbal replied:
added in latest commit

mhanson-github commented on src/index.ts:
```diff
@@ -5210,8 +5210,16 @@ const shardusSetup = (): void => {
+        if (isEIP2930 && (tooManyAddresses || tooManyStorageKeys)) {
+          return { 
+            status: false, 
+            reason: `EIP2930 tx blocked for having > ${ShardeumFlags.accessListSizeLimit} addresses in accessList or storage keys per address`
```
I'd make it less ambiguous and return a separate error return for each fail reason.

ahmxdiqbal replied:
ok

ahmxdiqbal committed:
clean-up eip2930 tx address-list/storage-keys check

ahmxdiqbal committed:
add unit tests for eip2930 tx check

mhanson-github committed:
Moved declaration of isEIP2930 closer to where it is used so it shows up in reviews.

Return after tooManyAddresses error so we don't do the expensive tooManyStorageKeys test if it has already failed from tooManyAddresses.

aniketdivekar commented on src/index.ts:
```diff
@@ -5202,16 +5202,30 @@ const shardusSetup = (): void => {
         let remoteTargetAccount
         appData.requestNewTimestamp = true // force all evm txs to generate a new timestamp
 
-        if (isEIP2930 && isStakingEVMTx(transaction) === false) {
```
we removed `isStakingEVMTx(transaction) === false` check altogether.. is that expected?
